### ISSUE TITLE
Sig(:final) does not work well in Static check

### ISSUE THREAD

**Issue by manhhung741** [sorbet/sorbet#1600]
#### Input

[→ View on sorbet.run](
https://sorbet.run/#%23%20typed%3A%20true%0A%0AT%3A%3AConfiguration.enable_final_checks_on_hooks%0A%0Amodule%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig(%3Afinal)%20%7B%20void%20%7D%0A%20%20def%20mmm%3B%20end%0A%0A%20%20def%20mmm%3B%20end%0Aend%0A%0Amodule%20B%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig(%3Afinal)%20%7B%20returns(NilClass)%20%7D%0A%20%20def%20mmm%3B%20end%0Aend%0A%0Aclass%20C%0A%20%20extend%20A%0A%20%20extend%20B%0A%0A%20%20def%20self.mmm%3B%20end%0Aend%0A%0AT.reveal_type(C.mmm)%0A
)

```ruby
# typed: true

T::Configuration.enable_final_checks_on_hooks

module A
  extend T::Sig

  sig(:final) { void }
  def mmm; end

  def mmm; end
end

module B
  extend T::Sig

  sig(:final) { returns(NilClass) }
  def mmm; end
end

class C
  extend A
  extend B

  def self.mmm; end
end

T.reveal_type(C.mmm)
```

#### Observed output

```
editor.rb:25: B#mmm was declared as final and cannot be overridden by C.mmm https://srb.help/5033
    25 |  def self.mmm; end
          ^^^^^^^^^^^^
    editor.rb:18: original method defined here
    18 |  def mmm; end
          ^^^^^^^

editor.rb:25: A#mmm was declared as final and cannot be overridden by C.mmm https://srb.help/5033
    25 |  def self.mmm; end
          ^^^^^^^^^^^^
    editor.rb:11: original method defined here
    11 |  def mmm; end
          ^^^^^^^

editor.rb:28: Revealed type: T.untyped https://srb.help/7014
    28 |T.reveal_type(C.mmm)
        ^^^^^^^^^^^^^^^^^^^^
  From:
    editor.rb:28:
    28 |T.reveal_type(C.mmm)
                      ^^^^^
Errors: 3
```

#### Expected behavior

- When `mmm` is redefined in module A, the static check should have shown error 
- When `mmm` is redefined in another module, the static check should have shown error

Currently it only show errors when a class reimplement a method from a module or superclass
- - -

**Comment by azdavis** [comment]
Thanks for reporting. Both of these issues are known issues that we decided to ship `final` with. (We alluded to this on the docs site for final, but we didn't say exactly which constructs aren't caught by the statics; we probably should have done that.)

**Comment by manhhung741** [comment]
Yeah, I think we should make it clear in the doc. It was confusing to me
because I tried to copy example from the doc and it doesn't work statically
